be/src/vec/exec/format/json/new_json_reader.cpp

```diff
@@ -1007,4 +1015,560 @@ Status NewJsonReader::_read_one_message(std::unique_ptr<uint8_t[]>* file_buf, si
+        return Status::OK();
+    }
+
+    if (!_parsed_jsonpaths.empty() && _strip_outer_array) {
+        _total_rows = _json_value.count_elements().value();
+        _next_row = 0;
+
+        if (_total_rows == 0) {
+            // meet an empty json array.
+            *is_empty_row = true;
+        }
+    }
+    return Status::OK();
+}
+Status NewJsonReader::_simdjson_parse_json_doc(size_t* size, bool* eof) {
+    // read a whole message
+    SCOPED_TIMER(_file_read_timer);
+    const uint8_t* json_str = nullptr;
+    std::unique_ptr<uint8_t[]> json_str_ptr;
+    if (_line_reader != nullptr) {
+        RETURN_IF_ERROR(_line_reader->read_line(&json_str, size, eof));
+    } else {
+        size_t length = 0;
+        RETURN_IF_ERROR(_read_one_message(&json_str_ptr, &length));
+        json_str = json_str_ptr.get();
+        *size = length;
+        if (length == 0) {
+            *eof = true;
+        }
+    }
+
+    _bytes_read_counter += *size;
+    if (*eof) {
+        return Status::OK();
+    }
+    memcpy(_simdjson_ondemand_padding_buffer, json_str, *size);
+    auto error = _ondemand_json_parser
+                         ->iterate(std::string_view(reinterpret_cast<const char*>(
+                                                            _simdjson_ondemand_padding_buffer),
+                                                    *size),
+                                   _padded_size)
+                         .get(_original_json_doc);
+    auto return_quality_error = [&](fmt::memory_buffer& error_msg,
+                                    const std::string& doc_info) -> Status {
+        RETURN_IF_ERROR(_state->append_error_msg_to_file(
+                [&]() -> std::string { return doc_info; },
+                [&]() -> std::string { return fmt::to_string(error_msg); }, _scanner_eof));
+        _counter->num_rows_filtered++;
+        if (*_scanner_eof) {
+            // Case A: if _scanner_eof is set to true in "append_error_msg_to_file", which means
+            // we meet enough invalid rows and the scanner should be stopped.
+            // So we set eof to true and return OK, the caller will stop the process as we meet the end of file.
+            *eof = true;
+            return Status::OK();
+        }
+        return Status::DataQualityError(fmt::to_string(error_msg));
+    };
+    if (error != simdjson::error_code::SUCCESS) {
+        fmt::memory_buffer error_msg;
+        fmt::format_to(error_msg, "Parse json data for JsonDoc failed. code: {}, error info: {}",
+                       error, simdjson::error_message(error));
+        return return_quality_error(error_msg, std::string((char*)json_str, *size));
+    }
+    try {
+        // set json root
+        // if it is an array at top level, then we should iterate the entire array in
+        // ::_simdjson_handle_flat_array_complex_json
+        if (_parsed_json_root.size() != 0 &&
+            _original_json_doc.type() == simdjson::ondemand::json_type::object) {
```

Comment:
If _parsed_json_root.size() != 0 but _original_json_doc.type() is not object, it seems to be an error.

Reply:
it's handled in _simdjson_handle_flat_array_complex_json